File: clients/windows/src/Vertex.Core/Protocol/AssignMessage.cs
using System.Text.Json.Serialization;

namespace Vertex.Core.Protocol;

/// <summary>
/// IP assignment response from exit, received on <c>vpn/{exit}/{name}/control</c>.
/// Field names match Go and Swift exactly.
/// </summary>
public sealed record AssignMessage(
    /// <summary>Assigned TUN IP address (e.g. <c>10.9.0.5</c>).</summary>
    [property: JsonPropertyName("ip")]   string  Ip,
    /// <summary>Subnet mask (e.g. <c>255.255.255.0</c>).</summary>
    [property: JsonPropertyName("mask"), JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)] string? Mask,
    /// <summary>Gateway / exit TUN IP (e.g. <c>10.9.0.1</c>).</summary>
    [property: JsonPropertyName("gw")]   string  Gw,
    /// <summary>Exit's X25519 DH public key (base64) for session key derivation.</summary>
    [property: JsonPropertyName("dh"),   JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)] string? Dh);

File: clients/shared/VertexCore/Sources/VertexCore/Protocol/AssignMessage.swift
import Foundation

/// IP assignment response from exit after join handshake.
///
/// Received on: `vpn/{exit}/{name}/control`
/// Field names match Go implementation.
public struct AssignMessage: Codable, Sendable {
    /// Assigned TUN IP address (e.g. "10.9.0.5")
    public let ip: String

    /// Subnet mask (e.g. "255.255.255.0")
    public let mask: String?

    /// Gateway/exit TUN IP (e.g. "10.9.0.1")
    public let gw: String

    /// Exit's X25519 DH public key (base64) for session key derivation
    public let dh: String?

    public init(ip: String, mask: String? = nil, gw: String, dh: String? = nil) {
        self.ip = ip
        self.mask = mask
        self.gw = gw
        self.dh = dh
    }
}
